File: tests/helpers/sqlite_store/alerts.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime

from tests.helpers.sqlite_store.base import SqliteStoreMixin
from tests.helpers.sqlite_store.schema import utcnow_iso


@dataclass(slots=True)
class StoredAlertState:
    market_id: str
    is_low: bool
    last_alert_at: datetime | None


class AlertStoreMixin(SqliteStoreMixin):

    def get_alert_state(self, market_id: str) -> StoredAlertState:
        row = self.conn.execute(
            "SELECT market_id, is_low, last_alert_at FROM alert_state WHERE market_id = ?",
            (market_id,),
        ).fetchone()
        if row is None:
            return StoredAlertState(market_id=market_id, is_low=False, last_alert_at=None)
        last_alert_at = (
            datetime.fromisoformat(row["last_alert_at"]) if row["last_alert_at"] else None
        )
        return StoredAlertState(
            market_id=row["market_id"],
            is_low=bool(row["is_low"]),
            last_alert_at=last_alert_at,
        )

    def upsert_alert_state(self, state: StoredAlertState) -> None:
        self.conn.execute(
            """
            INSERT INTO alert_state (market_id, is_low, last_alert_at, updated_at)
            VALUES (?, ?, ?, ?)
            ON CONFLICT(market_id) DO UPDATE SET
              is_low = excluded.is_low,
              last_alert_at = excluded.last_alert_at,
              updated_at = excluded.updated_at
            """,
            (
                state.market_id,
                1 if state.is_low else 0,
                state.last_alert_at.isoformat() if state.last_alert_at else None,
                utcnow_iso(),
            ),
        )
        self.conn.commit()

File: greenfloor-engine/src/storage/sqlite/alerts.rs
use rusqlite::params;

use crate::error::{SignerError, SignerResult};

use super::{utcnow_iso, SqliteStore};

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct StoredAlertState {
    pub market_id: String,
    pub is_low: bool,
    pub last_alert_at: Option<String>,
}

impl SqliteStore {
    /// Get alert state.
    ///
    /// # Errors
    ///
    /// Returns an error if the operation fails.
    pub fn get_alert_state(&self, market_id: &str) -> SignerResult<StoredAlertState> {
        let mut stmt = self
            .conn
            .prepare(
                "SELECT market_id, is_low, last_alert_at FROM alert_state WHERE market_id = ?1",
            )
            .map_err(|err| {
                SignerError::Other(format!("failed to prepare alert_state query: {err}"))
            })?;
        let mut rows = stmt
            .query(params![market_id])
            .map_err(|err| SignerError::Other(format!("failed to query alert_state: {err}")))?;
        let Some(row) = rows
            .next()
            .map_err(|err| SignerError::Other(format!("failed to read alert_state row: {err}")))?
        else {
            return Ok(StoredAlertState {
                market_id: market_id.to_string(),
                is_low: false,
                last_alert_at: None,
            });
        };
        let is_low: i64 = row
            .get(1)
            .map_err(|err| SignerError::Other(format!("failed to read is_low: {err}")))?;
        Ok(StoredAlertState {
            market_id: row
                .get(0)
                .map_err(|err| SignerError::Other(format!("failed to read market_id: {err}")))?,
            is_low: is_low != 0,
            last_alert_at: row.get(2).ok(),
        })
    }

    /// Upsert alert state.
    ///
    /// # Errors
    ///
    /// Returns an error if the operation fails.
    pub fn upsert_alert_state(&self, state: &StoredAlertState) -> SignerResult<()> {
        self.conn
            .execute(
                r"
                INSERT INTO alert_state (market_id, is_low, last_alert_at, updated_at)
                VALUES (?1, ?2, ?3, ?4)
                ON CONFLICT(market_id) DO UPDATE SET
                  is_low = excluded.is_low,
                  last_alert_at = excluded.last_alert_at,
                  updated_at = excluded.updated_at
                ",
                params![
                    state.market_id,
                    i64::from(state.is_low),
                    state.last_alert_at,
                    utcnow_iso(),
                ],
            )
            .map_err(|err| SignerError::Other(format!("failed to upsert alert_state: {err}")))?;
        Ok(())
    }
}
